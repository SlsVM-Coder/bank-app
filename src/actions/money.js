import { types } from '../types/types.js';
import { db } from '../firebase/firebaseConfig';
import { loadData } from './../helpers/loadData';

export const startLoadingData = (email) => {
  return async (dispatch) => {
    const data = await loadData(email);
    dispatch(dataLoaded(data));
  };
};

export const dataLoaded = (data) => ({
  type: types.dataLoaded,
  payload: data,
});

export const depositMoney = (quantity) => {
  return async (dispatch, getState) => {
    const { email, name } = getState().auth;
    const { balance } = getState().money;
    const newQuantity = balance + quantity;
    dispatch(startUpdateBalance(email, newQuantity));
    const movement = {
      sender: name,
      receiver: name,
      quantity,
      action: 'Deposit',
      date: Date.now(),
    };
    dispatch(startAddMovement(email, movement));
  };
};

export const transferMoney = (quantity, receiverEmail) => {
  return async (dispatch, getState) => {
    debugger;
    const { email, name: senderName } = getState().auth;
    const { balance } = getState().money;
    const snapReceiverInfo = await db.doc(`${receiverEmail}/info`).get();
    const { name: receiverName } = snapReceiverInfo.data();
    const newQuantity = balance - quantity;
    dispatch(startUpdateBalance(email, newQuantity)); // users update

    const movement = {
      sender: senderName,
      receiver: receiverName,
      quantity,
      action: 'Transfer',
      date: Date.now(),
    };
    dispatch(startAddMovement(email, movement));

    const snapReceiverBalance = await db.collection(`${receiverEmail}/`).get();
    const [{ balance: receiverBalance }] = snapReceiverBalance.docs.map((doc) =>
      doc.data(),
    );
    const newQuantityReceiver = receiverBalance + quantity;
    dispatch(startUpdateBalance(receiverEmail, newQuantityReceiver)); // receiver update

    const receiverMovement = {
      sender: senderName,
      receiver: receiverName,
      quantity,
      action: 'Deposit',
      date: Date.now(),
    };
    dispatch(startAddMovement(receiverEmail, receiverMovement));
  };
};

export const startUpdateBalance = (email, newQuantity) => {
  return async (dispatch, getState) => {
    await db.doc(`/${email}/data/`).update({ balance: newQuantity });

    const { email: userEmail } = getState().auth;

    if (userEmail === email) {
      dispatch(updateBalance(newQuantity));
    }
  };
};

export const updateBalance = (quantity) => ({
  type: types.updateBalance,
  payload: quantity,
});

export const startAddMovement = (email, movement) => {
  return async (dispatch) => {
    const doc = await db.collection(`${email}/data/movements`).add(movement);
    dispatch(addMovement(doc.id, movement));
  };
};

export const addMovement = (id, movement) => ({
  type: types.addMovement,
  payload: {
    id,
    ...movement,
  },
});

export const moneyLogout = () => ({
  type: types.moneyLogout,
});
