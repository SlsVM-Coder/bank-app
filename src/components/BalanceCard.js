import React from 'react';
import { useSelector } from 'react-redux';

export const BalanceCard = () => {
  const { balance } = useSelector((state) => state.money);
  return (
    <div className="BalanceCard">
      <h2>Balance:</h2>
      {balance && <p className="Number"> {balance}â‚¬</p>}
    </div>
  );
};
