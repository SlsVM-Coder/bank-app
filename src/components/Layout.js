import React from 'react';

export const Layout = ({ component, img, title, children, ...rest }) => {
  return (
    <>
      <div className={`Container ${component}`}>
        <div className="LeftContent">
          <div className="Content">
            {title && <h1> {`${title}`} </h1>}
            {children}
          </div>
        </div>
        <div className="RightContent">
          <img src={require(`./../assets/${img}.png`)} alt={img} />
        </div>
      </div>
    </>
  );
};
