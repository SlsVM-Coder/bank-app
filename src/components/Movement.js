import React from 'react';

export const Movement = ({
  receiver,
  quantity,
  sender,
  date,
  action,
}) => {
  return (
    <div className="Movement">
      <div className="Details">
        <h3>From: </h3>
        <span>{sender}</span>
        <h3>To: </h3>
        <span>{receiver} </span>
        <h3>Date:</h3>
        <span>{date}</span>
      </div>
      <div className="Quantity">
        <h2>{`${action === 'Transfer' ? '-' : '+'} ${quantity}â‚¬`}</h2>
      </div>
    </div>
  );
};
