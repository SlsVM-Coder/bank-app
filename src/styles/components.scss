.btn-primary {
  background: $black;
  padding: 8px 13px;
  color: white;
  font-size: 16px;
  border-radius: $border-radius;
  border: 1px solid transparent;
  cursor: pointer;
  transition: all 0.3s ease;

  &:focus {
    outline: 0;
  }

  &:hover {
    background-color: transparent;
    border: 1px solid $black;
    color: $black;
  }
}

.TheNavbar {
  height: 70px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  box-shadow: 0px 5px 30px rgba(25, 28, 31, 0.08);
  padding: 10px;
}

.input-text {
  color: $dark-grey;
  border: 1px solid $light-grey;
  border-radius: $border-radius;
  padding-left: 10px;
  margin-top: 8px;
  margin-bottom: 15px;
  font-size: 16px;
  height: 30px;
  width: 250px;
  transition: border-bottom 0.3s ease;

  &:focus {
    border: 1px solid $primary-color;
    outline: none;
  }

  &::placeholder {
    font-size: 14px;
  }
}

.BalanceCard {
  height: 170px;
  width: 90%;
  max-width: 350px;
  padding: 10px;
  border: 2px solid $black;
  border-radius: $border-radius;

  .Number {
    margin: 0;
    font-size: 50px;
  }
}

.Movement {
  border-bottom: 3px solid $black;
  padding-bottom: 10px;
  margin-bottom: 10px;
  display: flex;
  align-items: center;

  .Details {
    flex-grow: 1;
    display: grid;
    grid-template-columns: 70px 1fr;
    grid-template-rows: repeat(3, 22px);
    align-items: center;
  }

  .Quantity {
    flex-grow: 1;
    display: flex;
    justify-content: flex-end;
  }
}
