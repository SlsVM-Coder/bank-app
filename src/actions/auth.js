import Swal from 'sweetalert2';
import { types } from '../types/types.js';
import { db, firebase } from '../firebase/firebaseConfig';
import { startLoading, finishLoading } from './ui';
import { startLoadingData, moneyLogout } from './money';

export const startLoginEmailPassword = (email, password) => {
  return (dispatch) => {
    dispatch(startLoading());

    return firebase
      .auth()
      .signInWithEmailAndPassword(email, password)
      .then(({ user }) => {
        dispatch(startLoadingData(user.email));
        dispatch(login(user.uid, user.email, user.displayName));
        dispatch(finishLoading());
      })
      .catch((e) => {
        dispatch(finishLoading());
        Swal.fire('Error', e.message, 'error');
      });
  };
};
export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    return firebase
      .auth()
      .createUserWithEmailAndPassword(email, password)
      .then(async ({ user }) => {
        await user.updateProfile({ displayName: name });
        const info = {
          uid: user.uid,
          email: user.email,
          name: user.displayName,
        };
        await db.doc(`${user.email}/info`).set(info);
        await db.doc(`/${user.email}/data/`).set({ balance: 0 });
        dispatch(login(user.uid, user.email, user.displayName));
        dispatch(startLoadingData(user.email));
      })
      .catch((e) => {
        console.log(e);
        Swal.fire('Error', e.message, 'error');
      });
  };
};

export const login = (uid, email, displayName) => ({
  type: types.login,
  payload: {
    uid,
    email,
    displayName,
  },
});

export const startLogout = () => {
  return async (dispatch) => {
    await firebase.auth().signOut();
    dispatch(logout());
    dispatch(moneyLogout());
  };
};

export const logout = () => ({
  type: types.logout,
});
