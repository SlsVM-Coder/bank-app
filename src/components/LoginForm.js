import React from 'react';
import { useDispatch } from 'react-redux';
import { useForm } from './../hooks/useForm';
import { startLoginEmailPassword } from './../actions/auth';

export const LoginForm = () => {
  const dispatch = useDispatch();

  const [formValues, handleInputChange] = useForm({
    email: '',
    password: '',
  });

  const { email, password } = formValues;

  const handleLogin = (e) => {
    e.preventDefault();
    dispatch(startLoginEmailPassword(email, password));
  };

  return (
    <form
      className="animate__animated animate__fadeIn animate__faster"
      onSubmit={handleLogin}
    >
      <input
        type="text"
        placeholder="Email"
        name="email"
        className="input-text"
        autoComplete="off"
        value={email}
        onChange={handleInputChange}
      />

      <input
        type="password"
        placeholder="Password"
        name="password"
        className="input-text"
        value={password}
        onChange={handleInputChange}
      />

      <button type="submit" className="btn-primary">
        Sign in
      </button>
    </form>
  );
};
