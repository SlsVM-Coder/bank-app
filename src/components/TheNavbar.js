import React from 'react';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { startLogout } from '../actions/auth';

export const TheNavbar = () => {
  const dispatch = useDispatch();
  const { uid } = useSelector((state) => state.auth);

  const handleSignOut = () => {
    dispatch(startLogout());
  };

  return (
    <nav className="TheNavbar">
      <Link to="/" className="m-2">
        Smart Bank
      </Link>
      {uid ? (
        <button className="btn-primary m-2" onClick={handleSignOut}>
          <span>Sign out</span>
        </button>
      ) : (
        <Link to="/auth">
          <button renderas="button" className="btn-primary m-2">
            <span>Sign in</span>
          </button>
        </Link>
      )}
    </nav>
  );
};
