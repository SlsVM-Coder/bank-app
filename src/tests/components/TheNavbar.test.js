import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { MemoryRouter } from 'react-router-dom';
import { TheNavbar } from './../../components/TheNavbar';
import { startLogout } from '../../actions/auth';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
  auth: {
    name: 'Test name',
    email: 'test-bank@testing.com',
    uid: 'uid test',
  },
  money: {
    balance: 20,
    movements: [],
  },
};

jest.mock('../../actions/auth', () => ({
  startLogout: jest.fn(),
}));

let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter>
      <TheNavbar />
    </MemoryRouter>
  </Provider>,
);
describe('Pruebas en <TheNavbar />', () => {
  beforeEach(() => {
    store = mockStore(initState);
    jest.clearAllMocks();
  });

  test('debe de mostrarse correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('debe de llamar el logout', () => {
    wrapper.find('button').prop('onClick')();
    expect(startLogout).toHaveBeenCalled();
  });
});
