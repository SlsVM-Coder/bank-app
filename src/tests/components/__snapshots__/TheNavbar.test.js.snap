// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TheNavbar /> debe de mostrarse correctamente 1`] = `
<nav
  className="TheNavbar"
>
  <a
    className="m-2"
    href="/"
    onClick={[Function]}
  >
    Smart Bank
  </a>
  <button
    className="btn-primary m-2"
    onClick={[Function]}
  >
    <span>
      Sign out
    </span>
  </button>
</nav>
`;
